bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java

```diff
@@ -724,69 +730,63 @@ public Image(Device device, ImageFileNameProvider imageFileNameProvider) {
-			if (filename.zoom() == newZoom) {
-				/* Release current native resources */
-				destroy ();
-				initNative(filename.element());
-				if (this.handle == 0) init(new ImageData (filename.element()));
-				init();
-				refreshed = true;
-			} else {
-				/* Release current native resources */
-				destroy ();
-				ImageData resizedData = DPIUtil.autoScaleImageData (device, new ImageData (filename.element()), filename.zoom());
-				init(resizedData);
-				init ();
-				refreshed = true;
-			}
-			setCurrentNativeZoom(nativeZoom);
-		}
-	} else if (imageDataProvider != null) {
-		if (newZoom != oldZoom) {
-			ElementAtZoom<ImageData> data = DPIUtil.validateAndGetImageDataAtZoom (imageDataProvider, newZoom);
+public static Long win32_getHandle (Image image, int deviceZoomLevel) {
+	if(image.isDisposed()) {
+		return image.handle;
+	}
+	if(image.handleMap.get(deviceZoomLevel) != null) {
+		return image.handleMap.get(deviceZoomLevel);
+	}
+
+	if (image.imageFileNameProvider != null) {
+		ElementAtZoom<String> imageCandidate = DPIUtil.validateAndGetImagePathAtZoom (image.imageFileNameProvider, deviceZoomLevel);
+		if (imageCandidate.zoom() == deviceZoomLevel) {
 			/* Release current native resources */
-			destroy ();
-			ImageData resizedData = DPIUtil.autoScaleImageData (device, data.element(), data.zoom());
-			init(resizedData);
-			init();
-			refreshed = true;
-			setCurrentNativeZoom(nativeZoom);
+			long handle = image.initNative(imageCandidate.element(), deviceZoomLevel);
+			if (handle == 0) image.init(new ImageData (imageCandidate.element()), deviceZoomLevel);
+			image.init();
+		} else {
+			ImageData resizedData = DPIUtil.autoScaleImageData (image.device, new ImageData (imageCandidate.element()), deviceZoomLevel, imageCandidate.zoom());
+			image.init(resizedData, deviceZoomLevel);
+			image.init ();
 		}
-	} else if (this.dataAtBaseZoom != null) { // Resizing is only possible with a cached base image
-		if (newZoom != oldZoom) {
-			ImageData resizedData = getImageData(newZoom);
-			destroy ();
-			init(resizedData);
-			init();
-			refreshed = true;
-			setCurrentNativeZoom(nativeZoom);
+	} else if (image.imageDataProvider != null) {
+		ElementAtZoom<ImageData> imageCandidate = DPIUtil.validateAndGetImageDataAtZoom (image.imageDataProvider, deviceZoomLevel);
+		ImageData resizedData = DPIUtil.autoScaleImageData (image.device, imageCandidate.element(), deviceZoomLevel, imageCandidate.zoom());
+		image.init(resizedData, deviceZoomLevel);
+		image.init();
+	} else {
+		if (image.dataAtBaseZoom == null && image.memGC == null) {
+			// Cache data at base zoom before refresh.
+			image.dataAtBaseZoom = new ElementAtZoom<>(image.getImageData(image.getZoom()), image.getZoom());
+		}
+		if (image.dataAtBaseZoom != null) {
+			ImageData resizedData = image.getImageData(deviceZoomLevel);
+			image.init(resizedData, deviceZoomLevel);
+			image.init();
 		}
 	}
-	return refreshed;
+	return image.handleMap.get(deviceZoomLevel);
```

Comment:
Is there a scenario, where this could be null?

Reply:
Yes it is always checked in the block above and if not available, it is added.